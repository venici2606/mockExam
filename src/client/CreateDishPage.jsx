import React, { useState } from "react";
import { InputField } from "./InputField";

export function CreateDishPage() {
  const [name, setName] = useState("");
  const [price, setPrice] = useState("");

  async function submit(e) {
    e.preventDefault();
    await fetch("/api/dishes", {
      method: "POST",
      body: JSON.stringify({ name, price }),
      headers: {
        "Content-Type": "application/json",
      },
    });

    alert("You added a dish!");
  }

  return (
    <form onSubmit={submit}>
      <h1>Create a new dish</h1>
      <InputField label={"Dish name"} value={name} onChangeValue={setName} />
      <InputField
        label={"Price"}
        value={price}
        onChangeValue={setPrice}
        type={"number"}
      />
      <button>Submit</button>
    </form>
  );
}
