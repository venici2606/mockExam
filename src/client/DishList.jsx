import React from "react";
import { LoadingView } from "./components/LoadingView";
import { useLoading } from "./useLoading";
import { ErrorView } from "./components/ErrorView";

export function DishList({ dishApi }) {
  const { data: dishes, error, loading, reload } = useLoading(
    async () => await dishApi.listDishes()
  );

  if (error) {
    return <ErrorView error={error} reload={reload} />;
  }

  if (loading || !dishes) {
    return <LoadingView />;
  }

  return (
    <>
      <h3>
        <i>Menu</i>
      </h3>
      {dishes.map(({ id, name, price }) => (
        <li key={id}>
          Dish {id}: {name} {price} kr.
        </li>
      ))}
    </>
  );
}
