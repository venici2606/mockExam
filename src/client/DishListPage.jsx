import React from "react";
import { LoadingView } from "./components/LoadingView";
import { Link } from "react-router-dom";
import { useLoading } from "./useLoading";
import { ErrorView } from "./components/ErrorView";

export function DishListPage({ dishApi, loadProfile, msg }) {
  const { data: dishes, error, loading, reload } = useLoading(
    async () => await dishApi.listDishes()
  );

  const message = msg;
  const { data } = useLoading(async () => await loadProfile());

  if (error) {
    return <ErrorView error={error} reload={reload} />;
  }

  if (loading || !dishes) {
    return <LoadingView />;
  }

  /* If login add link */
  function output(data, name, id, price) {
    if (data)
      return (
        <Link to={`/dishes/${id}/edit`}>
          Dish {id}: {name} {price} kr
        </Link>
      );
    else return `Dish ${id}: ${name} ${price} kr`;
  }

  /*  return (
    <>
      <h1>Menu</h1>
      {dishes.map(({ id, name, price }) => (
        <li key={id}>
          {/!*          <Link to={`/dishes/${id}/edit`}>
            Dish {id}: {name} {price} kr
          </Link>*!/}
          {output(data, name, id, price)}
        </li>
      ))}
      <p>
        <i>You can click on the dishes to change them!</i>
      </p>
    </>
  );*/

  return (
    <div>
      <h3>Menu</h3>
      {dishes.map(({ id, name, price }) => (
        <li key={id}>{output(data, name, id, price)}</li>
      ))}
      <p>{message}</p>
    </div>
  );
}
