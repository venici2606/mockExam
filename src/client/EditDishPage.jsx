import React, { useState } from "react";
import { LoadingView } from "./components/LoadingView";
import { InputField } from "./components/InputField";
import { useLoading } from "./lib/useLoading";
import { ErrorView } from "./components/ErrorView";
import { useHistory, useParams } from "react-router";

function EditDishForm({ dish, onSubmit }) {
  const [name, setName] = useState(dish.name);
  const [price, setPrice] = useState(dish.price);
  const history = useHistory();

  async function submit(e) {
    onSubmit(e, { name, price });
    history.push("/dishes");

    alert("You changed the dish!");
  }

  return (
    <form onSubmit={submit}>
      <h1>Edit existing dish ({name})</h1>
      <InputField label={"Dish name"} value={name} onChangeValue={setName} />
      <InputField
        label={"Price"}
        value={price}
        onChangeValue={setPrice}
        type={"number"}
      />
      <button>Submit</button>
    </form>
  );
}

export function EditDishPage({ dishApi }) {
  const { id } = useParams();

  const { loading, error, data: dish, reload } = useLoading(
    async () => await dishApi.getDish(id),
    [id]
  );

  async function handleSubmit(e, { name, price }) {
    e.preventDefault();
    await dishApi.updateDish(id, { name, price });

    alert("You changed the dish!");
  }

  if (error) {
    return <ErrorView error={error} reload={reload} />;
  }

  if (loading || !dish) {
    return <LoadingView />;
  }

  return <EditDishForm dish={dish} onSubmit={handleSubmit} />;
}
