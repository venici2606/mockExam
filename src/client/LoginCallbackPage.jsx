import React, { useState, useEffect } from "react";
import { useHistory } from "react-router";

export function LoginCallbackPage({ onAccessToken }) {
  const hash = Object.fromEntries(
    new URLSearchParams(window.location.hash.substr(1))
  );

  const [error, setError] = useState();

  const history = useHistory();

  useEffect(() => {
    const loginState = JSON.parse(sessionStorage.getItem("loginState"));
    const { access_token, state, error } = hash;

    if (state !== loginState.state) {
      alert("Why are you here?");
      return;
    }
    if (error) {
      setError(error);
      return;
    }

    onAccessToken(access_token);
    //sessionStorage.removeItem("loginState");
    history.push("/");
  }, [hash]);

  if (error) {
    return (
      <div>
        <h1>An error occurred</h1>
        <div>{error.toString()}</div>
      </div>
    );
  }

  return <h1>Login callback</h1>;
}
