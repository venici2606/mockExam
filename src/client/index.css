#container {
    position: absolute;
    padding: 20px;
    background-color: #efeaf8;
    height: 500px;
    width: 500px;

    top: 0;
    bottom: 0;
    left: 0;
    right: 0;
    margin: auto;
}

h1, #lProfile, #lLogin, #btnLogin, .profile {
    text-align: center;
}

p {
    text-align: end;
}



