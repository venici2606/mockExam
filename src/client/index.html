<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <title>Cantine Webapp</title>
  </head>
  <body>
    <div id="app"></div>
  </body>
  <script src="index.jsx"></script>
</html>
