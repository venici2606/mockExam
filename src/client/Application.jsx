import React, { useEffect, useState } from "react";
import { BrowserRouter, Link } from "react-router-dom";
import { Route, Switch } from "react-router";
import { ProfilePage } from "./ProfilePage";
import { fetchJSON, fetchJson, postJSON } from "./lib/http";
import { LoginPage } from "./LoginPage";
import { LoginCallbackPage } from "./LoginCallbackPage";
import { CreateDishPage } from "./CreateDishPage";
import { DishListPage } from "./DishListPage";
import { EditDishPage } from "./EditDishPage";



function useLocalStorage(key) {
  const [value, setValue] = useState(() =>
    JSON.parse(localStorage.getItem(key))
  );

  useEffect(() => {
    if (value) {
      localStorage.setItem(key, JSON.stringify(value));
    } else {
      localStorage.removeItem(key);
    }
  }, [value, ]);
  return [value, setValue];
}

export function Application() {
  const dishApi = {
    listDishes: async () => await fetchJSON("/api/dishes"),
    getDish: async (id) => await fetchJSON(`/api/dishes/${id}`),
    createDish: async ({ name, price }) => {
      return postJSON("/api/dishes", {
        json: { name, price },
        method: "POST",
      });
    },
    updateDish: async (id, { name, price }) =>
      postJSON(`/api/dishes/${id}`, {
        json: { name, price },
        method: "PUT",
      }),
  };

  const [access_token, setAccess_token] = useLocalStorage("access_token");

  const googleIdentityProvider = {
    discoveryURL:
      "https://accounts.google.com/.well-known/openid-configuration",
    client_id:
      "186464366390-duk52tjp6srhuc4jdhh7mvhi07m1s40h.apps.googleusercontent.com",
    scope: "openid email profile",
  };

  async function loadProfile() {
    return fetchJson("/api/profile", {
      headers: {
        ...(access_token ? { Authorization: `Bearer ${access_token}` } : {}),
      },
    });
  }
  
  return (
    <div id={"container"}>
      <BrowserRouter>
        {/*Front page - Components*/}
        <Link to={"/"}>Front Page</Link>
        <Switch>
          <Route exact path={"/"} id={"container"}>
            <h1>Kristiania Cantine</h1>
            <Link to={"/profile"}>Profile</Link>
            <br/>
            <Link to={"/login"}>Login</Link>
            <DishListPage dishApi={dishApi} />
          </Route>

          {/*Login - Components*/}
          <Route path={"/profile"}>
            <ProfilePage loadProfile={loadProfile} />
          </Route>
          <Route exact path={"/login"}>
            <LoginPage identityProvider={googleIdentityProvider} />
          </Route>
          <Route path={"/login/callback"}>
            <LoginCallbackPage
              identityProvider={googleIdentityProvider}
              onAccessToken={(access_token) => setAccess_token(access_token)}
            />
          </Route>

          {/*Cantine - Components*/}
          <Route path={"/create"}>
            <CreateDishPage dishApi={dishApi} />
          </Route>
          <Route exact path={"/dishes"}>
            <DishListPage
              loadProfile={loadProfile}
              dishApi={dishApi}
              msg={"You can click on the dishes to change them!"}
            />
          </Route>
          <Route path={"/dishes/:id/edit"}>
            <EditDishPage dishApi={dishApi} />
          </Route>
          
          <Route>
            <h1>Page not found</h1>
          </Route>
        </Switch>
      </BrowserRouter>
    </div>
  );
}
