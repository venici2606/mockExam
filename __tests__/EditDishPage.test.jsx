import {EditDishPage} from "../src/client/EditDishPage";
import React from "react";
import ReactDOM from "react-dom";
import {MemoryRouter} from "react-router";
import {act} from "react-dom/test-utils";

describe("edit dish page", () => {
    it("can show information about an existing dish", async () => {
        const getDish = () => ({
            name: "Pizza",
            price: "119"
        });
        const container = document.createElement("div");
        await act(async () => {
            await ReactDOM.render(<MemoryRouter> <EditDishPage dishApi={{getDish }}/> </MemoryRouter>, container);
        });
        expect(container.innerHTML).toMatchSnapshot();
        expect(container.querySelector("h1").textContent).toEqual("Edit existing dish (Pizza)");
    });





    it("can show loading screen", async () => {
        const getDish = () => new Promise((resolve) => {});
        const container = document.createElement("div");
        await act(async () => {
            await ReactDOM.render(<MemoryRouter> <EditDishPage dishApi={{getDish }}/> </MemoryRouter>, container);
        });
        expect(container.innerHTML).toMatchSnapshot();
        expect(container.querySelector("div").textContent).toEqual("Loading ...");
    });

});