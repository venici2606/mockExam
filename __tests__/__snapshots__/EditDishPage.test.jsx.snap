// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`edit dish page can show information about an existing dish 1`] = `" <form><h1>Edit existing dish (Pizza)</h1><div><label>Dish name: <input type=\\"text\\" value=\\"Pizza\\"></label></div><div><label>Price: <input type=\\"number\\" value=\\"119\\"></label></div><button>Submit</button></form> "`;

exports[`edit dish page can show loading screen 1`] = `" <div>Loading ...</div> "`;
